src/rules/utils.ts

```diff
@@ -0,0 +1,685 @@
+  Node extends ExpectMember<Name> = ExpectMember<Name>
+> {
+  name: Name;
+  node: Node;
+}
+
+export interface ParsedExpectMatcher<
+  Matcher extends MatcherName = MatcherName,
+  Node extends ExpectMember<Matcher> = ExpectMember<Matcher>
+> extends ParsedExpectMember<Matcher, Node> {
+  arguments: TSESTree.CallExpression['arguments'] | null;
+  // isNamed<PossibleName extends Matcher>(
+  //   name: PossibleName,
+  //   ...or: PossibleName[]
+  // ): this is ParsedExpectMatcher<PossibleName>;
+  // isNamed<PossibleName extends Matcher>(
+  //   // name: PossibleName,
+  //   ...names: [PossibleName, ...PossibleName[]]
+  // ): this is ParsedExpectMatcher<PossibleName, Node>;
+}
+
+type BaseParsedModifier<
+  Modifier extends ModifierName = ModifierName
+> = ParsedExpectMember<Modifier>;
+
+type NegatableModifierName = ModifierName.rejects | ModifierName.resolves;
+type NotNegatableModifierName = ModifierName.not;
+
+interface NegatableParsedModifier<
+  Modifier extends NegatableModifierName = NegatableModifierName
+> extends BaseParsedModifier<Modifier> {
+  negation?: ExpectMember<ModifierName.not>;
+}
+
+export interface NotNegatableParsedModifier<
+  Modifier extends NotNegatableModifierName = NotNegatableModifierName
+> extends BaseParsedModifier<Modifier> {
+  negation?: never;
+}
+
+type ParsedExpectModifier =
+  | NotNegatableParsedModifier<NotNegatableModifierName>
+  | NegatableParsedModifier<NegatableModifierName>;
+
+interface Expectation<ExpectNode extends ExpectCall = ValidExpectCall> {
+  expect: ExpectNode;
+  /** @deprecated */
+  expectCall: ExpectNode;
+  modifier?: ParsedExpectModifier;
+  matcher?: ParsedExpectMatcher;
+}
+
+const parseExpectMember = <S extends ExpectPropertyName>(
+  expectMember: ExpectMember<S>,
+): ParsedExpectMember<S> => ({
+  name: getAccessorValue<S>(expectMember.property),
+  node: expectMember,
+});
+
+const reparseAsMatcher = (
+  parsedMember: ParsedExpectMember,
+): ParsedExpectMatcher => ({
+  ...parsedMember,
+  /**
+   * The arguments being passed to this `Matcher`, if any.
+   *
+   * If this matcher isn't called, this will be `null`.
+   */
+  arguments:
+    parsedMember.node.parent &&
+    parsedMember.node.parent.type === AST_NODE_TYPES.CallExpression
+      ? parsedMember.node.parent.arguments
+      : null,
+  // isNamed(...names: string[]) {
+  //   return names.includes(this.name);
+  // },
+});
+
+/**
+ * Re-parses the given `parsedMember` as a `ParsedExpectModifier`.
+ *
+ * If the given `parsedMember` does not have a `name` of a valid `Modifier`,
+ * an exception will be thrown.
+ *
+ * @param {ParsedExpectMember<ModifierName>} parsedMember
+ *
+ * @return {ParsedExpectModifier}
+ */
+const reparseMemberAsModifier = (
+  parsedMember: ParsedExpectMember<ModifierName>,
+): ParsedExpectModifier => {
+  if (isSpecificMember(parsedMember, ModifierName.not)) {
+    return parsedMember;
+  }
+
+  /* istanbul ignore next */
+  if (
+    !(
```

Comment:
way too many parens

```ts
if (
  !isSpecificMember(parsedMember, ModifierName.resolves) &&
  !isSpecificMember(parsedMember, ModifierName.rejects)
) {
  throw
}
```

Maybe even `isSpecificMember` should take varargs at the end so you can do `isSpecificMember(parsedMember, ModifierName.resolves, ModifierName.rejects)`?

Reply:
Agreed - I'm not sure I went w/ that in the first place :joy:

While varargs would work, I didn't use them b/c I think it's not worth it - this is the only place that method is used, and w/ using varargs you've got to `& string[]` for `.includes` to work properly; I don't think the mess is worth the gain here.

----

Overall the whole reason we need this check is that TypeScript knows that while compile time we enforce `name` & `getAccessorValue(node)` to be the "same" (since `name` is meant to be literally the result of `getAccessorValue(node)`), at runtime there is no such enforcement.

It's a bit of a messy situation, that I think is partly due to how I've structured the typings to account for negation, but imo it's an acceptable trade-off: this is as unsound as we get relative to the other types, and they should be as stable as they can get.

The alternative types would likely be very confusing & drawn out, for the same result minus having to do this check at runtime.

I think this will be solvable (or at least improvable) once negated types land, of which this library is one I've got earmarked as a nice place to play around w/ them in b/c there's some clear places they could be used, so hopefully they'll land soon.